File: python/tsuki_ux/_spinner.py
"""
tsuki_ux._spinner
~~~~~~~~~~~~~~~~~
Standalone spinner for tasks that don't stream line output.
Port of ui.go's Spinner type.

Usage:
    s = Spinner("Detectando puerto…")
    s.start()
    time.sleep(2)
    s.stop(ok=True, msg="Puerto detectado: /dev/ttyUSB0")

    # Custom frames:
    from tsuki_ux import SPINNER_FRAMES_MOON
    s = Spinner("Compilando…", frames=SPINNER_FRAMES_MOON)
    s.start()
    ...
    s.stop(ok=True)

    # Context manager:
    with Spinner("Instalando paquetes…") as s:
        time.sleep(1)
"""

from __future__ import annotations
import itertools
import sys
import threading
import time
from typing import List, Optional

from tsuki_ux._color import C_INFO, C_SUCCESS, C_ERROR, DIM, RESET
from tsuki_ux._symbols import SPINNER_FRAMES, SYM_OK, SYM_FAIL, SYM_ELL


def _is_tty() -> bool:
    return hasattr(sys.stdout, "isatty") and sys.stdout.isatty()


class Spinner:
    """Animated spinner — TTY-safe, non-blocking, supports custom frames."""

    _INTERVAL = 0.08  # seconds (≈ 12 fps — matches ui.go's 80 ms)

    def __init__(self, msg: str, frames: Optional[List[str]] = None) -> None:
        self.msg = msg
        self._frames = frames if frames is not None else SPINNER_FRAMES
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._tty = _is_tty()
        self._lock = threading.Lock()
        self._start_time: Optional[float] = None

    # ── Internal ───────────────────────────────────────────────────────────────

    def _fmt_elapsed(self) -> str:
        if self._start_time is None:
            return ""
        s = time.monotonic() - self._start_time
        if s < 1.0:
            return f"{int(s * 1000)}ms"
        return f"{s:.1f}s"

    # ── Public API ─────────────────────────────────────────────────────────────

    def start(self) -> None:
        """Start the spinner animation (non-blocking)."""
        self._start_time = time.monotonic()
        if not self._tty:
            sys.stdout.write(f"  {DIM}{SYM_ELL}{RESET}  {self.msg}\n")
            sys.stdout.flush()
            return

        sys.stdout.write(f"\033[?25l  {C_INFO}{self._frames[0]}{RESET}  {self.msg}")
        sys.stdout.flush()

        def _run() -> None:
            for frame in itertools.cycle(self._frames):
                if self._stop_event.is_set():
                    break
                with self._lock:
                    sys.stdout.write(f"\r  {C_INFO}{frame}{RESET}  {self.msg}")
                    sys.stdout.flush()
                self._stop_event.wait(self._INTERVAL)

        self._thread = threading.Thread(target=_run, daemon=True)
        self._thread.start()

    def update_label(self, msg: str) -> None:
        """Change the spinner label mid-animation."""
        with self._lock:
            self.msg = msg

    def stop(self, ok: bool = True, msg: str = "") -> None:
        """Stop the spinner and print the final status line.

        ok=True  →  ✔  label  [1.2s]
        ok=False →  ✖  label  reason
        """
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=self._INTERVAL * 3)

        elapsed = self._fmt_elapsed()

        if self._tty:
            sys.stdout.write("\r\033[K\033[?25h")

        final = msg or self.msg
        if ok:
            timing = f"  {DIM}[{elapsed}]{RESET}" if elapsed else ""
            sys.stdout.write(f"  {C_SUCCESS}{SYM_OK}{RESET}  {final}{timing}\n")
        else:
            reason = msg or "failed"
            sys.stdout.write(f"  {C_ERROR}{SYM_FAIL}{RESET}  {self.msg}  {DIM}{reason}{RESET}\n")
        sys.stdout.flush()

    def stop_silent(self) -> None:
        """Stop and erase without printing anything — used by LiveBlock."""
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=self._INTERVAL * 3)
        if self._tty:
            sys.stdout.write("\r\033[K\033[?25h")
            sys.stdout.flush()

    # ── Context manager ────────────────────────────────────────────────────────

    def __enter__(self) -> "Spinner":
        self.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> bool:
        self.stop(ok=exc_type is None, msg=str(exc_val) if exc_val else "")
        return False

File: rust/src/spinner.rs
//! Standalone `Spinner`. Port of ui.go's `Spinner` type.
//!
//! Supports custom frame sets via [`Spinner::with_frames`].

use std::io::Write;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc, Mutex,
};
use std::thread;
use std::time::{Duration, Instant};

use crate::color::{ansi, C_INFO, C_SUCCESS, C_ERROR, DIM, RESET};
use crate::symbols::{spinner_frames, sym_ok, sym_fail, sym_ell};

const INTERVAL: Duration = Duration::from_millis(80);

/// Animated spinner — TTY-safe, non-blocking, supports custom frame sets.
///
/// # Examples
///
/// ```no_run
/// use tsuki_ux::{Spinner, spinner_frames_moon};
///
/// // Default braille frames
/// let mut s = Spinner::new("Detectando puerto…");
/// s.start();
/// std::thread::sleep(std::time::Duration::from_secs(1));
/// s.stop(true, Some("Puerto: /dev/ttyUSB0"));
///
/// // Custom frames
/// let mut s = Spinner::with_frames("Compilando…", spinner_frames_moon());
/// s.start();
/// std::thread::sleep(std::time::Duration::from_secs(1));
/// s.stop(true, None);
/// ```
pub struct Spinner {
    msg:    Arc<Mutex<String>>,
    frames: &'static [&'static str],
    tty:    bool,
    stop:   Arc<AtomicBool>,
    handle: Option<thread::JoinHandle<()>>,
    start:  Instant,
}

impl Spinner {
    /// Create a spinner with the default braille frames.
    pub fn new(msg: &str) -> Self {
        Self::with_frames(msg, spinner_frames())
    }

    /// Create a spinner with custom animation frames.
    ///
    /// Use one of the `spinner_frames_*()` functions from [`crate::symbols`],
    /// or supply your own `&'static [&'static str]`.
    pub fn with_frames(msg: &str, frames: &'static [&'static str]) -> Self {
        Self {
            msg:    Arc::new(Mutex::new(msg.to_owned())),
            frames,
            tty:    crate::color::is_tty(),
            stop:   Arc::new(AtomicBool::new(false)),
            handle: None,
            start:  Instant::now(),
        }
    }

    /// Start animation (non-blocking).
    pub fn start(&mut self) {
        if !self.tty {
            println!("  {}{}{}  {}", ansi(DIM), sym_ell(), ansi(RESET),
                     self.msg.lock().unwrap());
            return;
        }
        print!("\x1b[?25l  {}  {}", C_INFO.paint(self.frames[0]),
               self.msg.lock().unwrap());
        std::io::stdout().flush().ok();

        let stop   = Arc::clone(&self.stop);
        let msg    = Arc::clone(&self.msg);
        let frames = self.frames;

        self.handle = Some(thread::spawn(move || {
            let mut i = 1usize;
            while !stop.load(Ordering::Relaxed) {
                thread::sleep(INTERVAL);
                let frame   = frames[i % frames.len()];
                let current = msg.lock().unwrap().clone();
                print!("\r  {}  {}", C_INFO.paint(frame), current);
                std::io::stdout().flush().ok();
                i += 1;
            }
        }));
    }

    /// Change the spinner label mid-animation.
    pub fn update_label(&self, label: &str) {
        if let Ok(mut m) = self.msg.lock() {
            *m = label.to_owned();
        }
    }

    /// Stop and print the final status line.
    ///
    /// ```text
    /// ok=true  →  ✔  label  [1.2s]
    /// ok=false →  ✖  label  reason
    /// ```
    pub fn stop(mut self, ok: bool, msg: Option<&str>) {
        self.stop.store(true, Ordering::Relaxed);
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }

        let elapsed = {
            let d = self.start.elapsed();
            if d.as_millis() < 1000 {
                format!("{}ms", d.as_millis())
            } else {
                format!("{:.1}s", d.as_secs_f64())
            }
        };

        if self.tty {
            print!("\r\x1b[K\x1b[?25h");
        }

        let label = self.msg.lock().unwrap().clone();
        if ok {
            let suffix = msg.map(|m| format!("  {}", m)).unwrap_or_default();
            println!("  {}  {}  {}[{}]{}{}",
                C_SUCCESS.paint(sym_ok()), label,
                ansi(DIM), elapsed, ansi(RESET),
                suffix);
        } else {
            let reason = msg.unwrap_or("failed");
            println!("  {}  {}  {}{}{}",
                C_ERROR.paint(sym_fail()), label,
                ansi(DIM), reason, ansi(RESET));
        }
        std::io::stdout().flush().ok();
    }

    /// Stop silently (success, no output).
    pub fn stop_silent(mut self) {
        self.stop.store(true, Ordering::Relaxed);
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
        if self.tty {
            print!("\r\x1b[K\x1b[?25h");
            std::io::stdout().flush().ok();
        }
    }
}